### PR TITLE
sys: net: introduce low-level ethernet driver API

### PR DESCRIPTION
**Pull request by kaspar030**
While implementing the ethernet driver for encx24j600, I realized that ng_netdev is too fat for some applications, so I tried to create an interface just for ethernet drivers.

The actual tap device handling is taken from #2558.

*edit*
Let me rephrase the "too fat" statement, as it is totally inaccurate.

IMHO the currently developed network stack might not be the only one we might have in RIOT.
So when I started developing an ethernet driver for RIOT, I didn't want to tie it to netdev and friends.
netdev does make some assumptions on how a driver has to be implemented, and also implies a lot of dependencies.

This PR introduces a fairly thin interface with no dependencies, which can easily be used with netdev (see #2776), but doesn't depend on it.

I've implemented two drivers using it (the tap driver in this PR and a driver for encx24j600 chips, which I'll PR soon). As always with interfaces designed by yourself, I'm quite happy with the result.

The only disadvantage of using this interface instead of directly writing a driver for netdev is that pktsnips have to be converted to a flat buffer before sending to the device. This could be added to the interface if deemed necessary.

### CODE REVIEW

File: cpu/native/ng_net/dev_eth_tap.c [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    }
+    else if (nread == -1) {
+        if ((errno == EAGAIN) || (errno == EWOULDBLOCK)) {
+            //warn("read");
+        }
+        else {
+            err(EXIT_FAILURE, "dev_eth_tap: read");
+        }
+    }
+    else {
```
Comment (cgundogan) [inline cpu/native/ng_net/dev_eth_tap.c]: Please have a look at #2686 and #2664 (for the case `nread == 0`)

**Comment by kaspar030** [comment]
@cgundogan integrated.

File: tests/dev_eth/main.c [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+extern dev_eth_tap_t dev_eth_tap;
+dev_eth_t *dev = (dev_eth_t*)&dev_eth_tap;
+#endif
+
+void dev_eth_isr(dev_eth_t *dev) {
+    (void)dev;
+    thread_wakeup(handler_pid);
+}
+
+void dev_eth_rx_handler(dev_eth_t *dev) {
```
Comment (haukepetersen) [inline tests/dev_eth/main.c]: this is never being called?

Comment (kaspar030) [inline tests/dev_eth/main.c]: It is supposed to be called from a drivers ```_isr``` implementation. Does the test app work (e.g., print something if you sent a packet to the node / broadcast one)?

**Comment by kaspar030** [comment]
- updated
- rebased

**Comment by haukepetersen** [comment]
I might have an idea to prevent having to copy the data to send into one continuous chunk before you send it: how about splitting the send function in three parts: `prepare`, `load`, and `execute`. This way you could call the `load` function with an data offset parameter multiple times and then call `exec` once all the data is loaded into the device buffer...  The original send function could stay for convenience reasons, it would just call the three 'lower-level' functions.

**Comment by haukepetersen** [comment]
Also I am not quite sure about the global functions you defined outside the `eth_driver_t` struct. I have the impression that they do not scale very well if you have multiple ethernet devices?!

**Comment by kaspar030** [comment]
> ... splitting the send function in three parts ...
I came to the same conclusion. Will change the interface accordingly.

> Also I am not quite sure about the global functions you defined outside the eth_driver_t struct. I have the impression that they do not scale very well if you have multiple ethernet devices?!

Why not? Those functions are called with the device handle as parameter. Should be easy to dispatch to whatever the network stack needs.

**Comment by kaspar030** [comment]
> load function with an data offset parameter

Do you think we can omit that parameter? The two drivers I have don't need it.

**Comment by OlegHahm** [comment]
I didn't really follow the latest discussions on netdev and this stuff, but looking at the proposed interface that looks exactly how I remember netdev to look like. Now, that I look at `driver/include/netdev` again, I see that it became a little bit bigger over time, but still I don't really get the purpose/advantage of your interface. What dependencies are pulled in by netdev?

**Comment by kaspar030** [comment]
@OlegHahm ng_netdev pulls at least ng_pkt, ng_pktbuf, ng_netconf. All of them are not needed for a low-level driver. pktbuf alone disqualifies the interface for class-0 devices. And as you realized, the interface is nearly the same, just without those dependencies. That is exactly the purpose / advantage of it.

**Comment by OlegHahm** [comment]
> ng_netdev pulls at least ng_pkt, ng_pktbuf, ng_netconf.

Maybe I'm starring at the wrong files. Can you point me to these dependencies?

> And as you realized, the interface is nearly the same, just without those dependencies. That is exactly the purpose / advantage of it.

Sounds completely reasonable to me. But this sounds to me as we should use this interface as a base and put additional features as a kind of add-on on top of it if necessary.


(Sorry if I discuss about this from a rather abstract level, but I don't have time to dig into the actual code too much.)

**Comment by kaspar030** [comment]
>>    ng_netdev pulls at least ng_pkt, ng_pktbuf, ng_netconf.

>Maybe I'm starring at the wrong files. Can you point me to these dependencies?

From ```ng_netdev.h```:
```
#include "net/ng_pkt.h"
#include "net/ng_netconf.h"
```

Actually pktbuf is not a direct dependency, but ```ng_pkt.h``` states:
```@note    This type has no initializer on purpose. Please use @ref net_ng_pktbuf as factory.```

Also the ```pktsnip_t``` is more complex than neccessary, forcing this additional complexity into every triver that uses ng_netdev.

> Sounds completely reasonable to me. But this sounds to me as we should use this interface as a base and put additional features as a kind of add-on on top of it if necessary.

Which interface is "this" in that sentence? :)

**Comment by OlegHahm** [comment]
Ah, I was indeed looking at the wrong file. Why is device driver interface specified in `sys`? Hm, `ng_netconf.h` is just header including some enums, so this dependency is bearable.

For the the other depencies, I agree that this is a bit unfortunate. Can we find a way to abstract from this?

I agree, that it may make sense to have different network stacks with different focuses (class 0 devices or class 1+ devices, for instance), but on the device driver level, I would have assumed that it is possible to define a basic set of functions that can (almost) always be satisfied.

> Which interface is "this" in that sentence? :)

Sorry, I meant the interface you propose in this PR.

**Comment by kaspar030** [comment]
>  Why is device driver interface specified in sys?

You're saying it should move to drivers/include?

> Hm, ng_netconf.h is just header including some enums, so this dependency is bearable.

Agreed. Still personally I always prefer not to dictate implementation through interface, just for the flexibility.

> I agree, that it may make sense to have different network stacks with different focuses (class 0 devices or class 1+ devices, for instance), but on the device driver level, I would have assumed that it is possible to define a basic set of functions that can (almost) always be satisfied.

Totally. That's what I tried with this interface. Should be simple to extend it so it also works for radio devices. And it can already be used with the ng network stack.

**Comment by OlegHahm** [comment]
> You're saying it should move to drivers/include?

That seems the natural place for me to look for a device driver interface - and that's the place where I searched for it.

> Totally. That's what I tried with this interface. Should be simple to extend it so it also works for radio devices. And it can already be used with the ng network stack.

@haukepetersen, @authmillenon, do you have any opinion on this?

**Comment by authmillenon** [comment]
@OlegHahm I'm also of the opinion to put it into drivers (as the legacy version of netdev you found there indicates). @haukepetersen however put it into `sys/include/net` since he feels it more belonging to the network architecture than the device architecture. The arguments for and against `drivers/include` or `sys/include/net` weigh each other pretty much out, I guess.

His argument was more or less, that yes, it's device related and it presents an API for a driver (that can in theory be directly connected to the hardware), but things like this PR and previous discussions not to enforce `ng_netdev` on devices, mainly because its strong dependency to `ng_pktbuf`, show that it is more an API for the network stack to speak with a device.

Does this reflect your argument @haukepetersen?

**Comment by OlegHahm** [comment]
Why is it that dependent on `ng_pktbuf`? In the end the device just needs a pointer to some data (for sending) or a pointer to some buffer where to copy data (for receiving).

IMO we should have network device interface (i.e. `netdev`) in `drivers/include` that can be used as a basis for all network stacks. Otherwise (being dependent on RIOT specific structures) we would have to implement a device driver for each network stack like OpenWSN.

**Comment by authmillenon** [comment]
@OlegHahm because a network device is required to at least release a sending packet from packet buffer, because it is the last one touching it. Apart from it sure, you can just use arbitrary `ng_pktsnip_t` pointers, but in practice it makes most sense to put them into the packet buffer directly to reduce overhead.

**Comment by haukepetersen** [comment]
I spend some more thoughts about these driver interfaces the last days, but I have not come up with a clear opinion, yet. I understand @kaspar030s thoughts about `netdev` being too complicated/having too many implications on the actual implementation, but I don't see them necessarily as too bad... 

I guess creating true low-level driver interfaces for ethernet, ieee802.15.4 etc and mapping the `netdev` interface on top of these would be an option. This is in my opinion however only valid, if we do not introduce addional copying of data and/or addiontial depth in function call hierarchy... Otherwise we would take efficiency away from the `netdev` interface.

**Comment by authmillenon** [comment]
My question about your opinion was more about the directory of `ng_netdev.h`.

**Comment by OlegHahm** [comment]
> This is in my opinion however only valid, if we do not introduce addional copying of data and/or addiontial depth in function call hierarchy...

Absolutely!

**Comment by OlegHahm** [comment]
> @OlegHahm because a network device is required to at least release a sending packet from packet buffer, because it is the last one touching it. 

Sure, but the releasing itself does not has do be done directly in the device driver.

> Apart from it sure, you can just use arbitrary ng_pktsnip_t pointers, but in practice it makes most sense to put them into the packet buffer directly to reduce overhead.

You're talking about receiving, I assume? What is different for filling the packet buffer compared to copying into an arbitrary buffer? (For the record: that's a serious question. As I stated previously, I did not follow all discussions and concepts for the lower layers.)

**Comment by kaspar030** [comment]
> What is different for filling the packet buffer compared to copying into an arbitrary buffer? (For the record: that's a serious question. As I stated previously, I did not follow all discussions and concepts for the lower layers.)

The dependency on packet buffer for one.

**Comment by authmillenon** [comment]
Needs rebase

**Comment by kaspar030** [comment]
- rebased

File: sys/include/net/dev_eth.h [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+/**
+ * @brief dev_eth event callback for packets that were received.
+ *
+ * This function should be called from whithin your driver's isr()
+ * (as defined in eth_driver_t), once for every packet that was received.
+ *
+ * It needs to call dev->driver->recv() in order to get received packet
+ * from the driver.
+ */
+extern void dev_eth_rx_handler(dev_eth_t *dev);
```
Comment (haukepetersen) [inline sys/include/net/dev_eth.h]: why this (and the `dev_eth_linkstate_handler()`) functions as extra functions? Aren't they just some specialized cases of 'dev_eth_isr()`?

**Comment by haukepetersen** [comment]
In general I like this PR. But I think there are some things that need to be changed. You claim `In order to be universally usable, it leaves out many implementation details to the implementation of an actual network stack using this interface.`. As this might be true for the driver implementation, this is not true for the stuff put on top. 
For example: To use this driver, you need to have all your data that you want to send in one continuous memory location. This is not true e.g. for the ng_ stack. 
So I would suggest to split the `send()` function to something like `prepare()`, `load(data, len, offset)` and `send()` or similar. Same goes for receiving. What do you think?

File: sys/include/net/dev_eth.h (hunk 3 above)
Comment (kaspar030) [inline sys/include/net/dev_eth.h]: The 'dev_eth_isr()' function lets the driver loop know that the ISR was triggered and there's work to do. It is called from interrupt context, and lets a thread know that driver->isr has to be run.

This driver->isr then calls dev_eth_rx_handler or the linkstate handler.

**Comment by kaspar030** [comment]
> So I would suggest to split the send() function to something like prepare(), load(data, len, offset) and send() or similar. Same goes for receiving. What do you think?

Yes. Will change it.

**Comment by kaspar030** [comment]
> So I would suggest to split the send() function to something like prepare(), load(data, len, offset) and send() or similar. Same goes for receiving. What do you think?

After thinking a little, how do we deal with devices which cannot implement this without a buffer in between? Right now only native tap comes to mind.

For receiving, does it make sense there?

**Comment by OlegHahm** [comment]
> After thinking a little, how do we deal with devices which cannot implement this without a buffer in between? Right now only native tap comes to mind.

You mean only native tap needs to buffer? I cannot think of any reasonable transceiver that does not provide a buffer (FIFO etc) in hardware.

**Comment by kaspar030** [comment]
> You mean only native tap needs to buffer? I cannot think of any reasonable transceiver that does not provide a buffer (FIFO etc) in hardware.

Yeah, all of them provide a buffer, but some might not support the "prepare, load*n, go" scheme. Well, let's cross that bridge when we arrive there. :)

**Comment by kaspar030** [comment]
@haukepetersen Do you think this is needed for reception?

**Comment by OlegHahm** [comment]
I think if the scheme is not supported (without additional code), it's perfectly fine just not to implement these features and provide some way to query the feature set (at compile or run time).

**Comment by kaspar030** [comment]
> provide some way to query the feature set (at compile or run time)

I'll try to find a way...

**Comment by kaspar030** [comment]
How do we handle locking? e.g., a single "send" locks, sends, unlocks a device.
Would it be enough if prepare locks and the final execute unlocks the device?

**Comment by haukepetersen** [comment]
Yes, I do think so!

**Comment by haukepetersen** [comment]
Regarding locking: I say the drivers are not supposed to be thread safe. It should be up to the user of the driver to make sure it is used in a coherent way. If thread safe access to the driver, I say an additional layer on top of the driver could be introduced by anyone needing this (or do it as in the new network stack where only one thread accesses the driver directly).

**Comment by kaspar030** [comment]
> Regarding locking: I say the drivers are not supposed to be thread safe.

Well, most drivers will lock the device during e.g., a send call anyways. We could just per convention make '_prepare' do the locking, '_load' expect a locked device and '_execute' unlock after sending.

**Comment by kaspar030** [comment]
> Yes, I do think so!

Do you have a use-case? The only that I can envision is that a stack could look at the headers of a packet, and then only read it's payload if necessary. But this would a lot of complexity, just to save the SPI read. At 5MHz SPI speed, an MCU can read ~30k 802.15.4 packets per second...

**Comment by kaspar030** [comment]
Do we expect c99-capable compilers?

If yes, we could go for a send function that takes a standard "struct iovec" as argument. As c99 supports dynamically sized arrays on stack (e.g., "struct iovec iov[num]" with num being a variable), it would be very easy to convert a pktsnip. native could use "writev" so it doesn't need an extra packet buffer to support the prepare/load/execute scheme, and it would reduce send to one function.

**Comment by OlegHahm** [comment]
RIOT requires ANSI C99 compliance from the compiler.

**Comment by kaspar030** [comment]
Nice.

Is anything using the prepare/load/execute scheme other than for getting pktsnips into a device?

**Comment by OlegHahm** [comment]
One main reasons for this scheme was to port the OpenWSN stack to this interface. (Currently, it's still using its own radio drivers.)

**Comment by kaspar030** [comment]
What are the requirements there?

I quickly skimmed the default stack. Looks like radio device are either in PRELOADING mode, which allows them to store a packet and send it later (when going into TX mode), or in a direct mode where a packet get's send at once.

Still, a packet is basically always copied into the device in one piece if I read the code right.

**Comment by kaspar030** [comment]
Alright, for sending, I think I found a solution for a send function that works fine in all our use cases.

1. we change the signature from "send(dev, buf, len)" to "send(dev, iovec, iovec_num)"
This allows easy partial loading (e.g., from pkt_snips).
2. if pktbuf is actually used, we #ifdef another send function: "send_pkt(dev, pktsnip)".
Internally, if a driver implements any of those in a sane way, both 1. and 2. can use the same internal functions for prepare/load/execute.
Also, to save implementation time, we can write a generic wrapper for send_pkt() -> send() once, and drivers can use that if they feel to do so.

If the device is not supposed to send the packet at once, that is handled using an internal PRELOAD state like in current netdev radios.

For reception, I had a chat with Oleg yesterday, and we fail to see a need for not reading a whole packet on reception. @haukepetersen, could you elaborate?

**Comment by OlegHahm** [comment]
@haukepetersen, ping.

**Comment by kaspar030** [comment]
@OlegHahm I've talked to Hauke, we have some design changes in mind.
@haukepetersen I'd still propose to get this in so the new network stack can more easily be tested on native, and improve from there. Ok?

**Comment by OlegHahm** [comment]
Ok, I agree that it would be very helpful to get this, #2776, and #2901 merged soonish to have an easy way to test the new stack on native.

**Comment by OlegHahm** [comment]
@kaspar030, please rebase. WIP label can also be removed, right?

**Comment by kaspar030** [comment]
- rebased

this should be adapted to the new netif auto-init scheme, but let's do that later...

File: cpu/native/ng_net/dev_eth_tap.c [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for
+ * more details.
+ */
+
+/*
+ * @ingroup net_dev_eth
+ * @{
+ * @file    dev_eth_tap.c
```
Comment (OlegHahm) [inline cpu/native/ng_net/dev_eth_tap.c]: remove the filename

File: cpu/native/ng_net/dev_eth_tap.c [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+        if (real_select(dev->tap_fd + 1, &rfds, NULL, NULL, &t) == 1) {
+            int sig = SIGIO;
+            extern int _sig_pipefd[2];
+            extern ssize_t (*real_write)(int fd, const void * buf, size_t count);
+            real_write(_sig_pipefd[1], &sig, sizeof(int));
+            _native_sigpend++;
+            DEBUG("dev_eth_tap: sigpend++\n");
+        }
+        else {
+            //DEBUG("dev_eth_tap: no more pending tap data\n");
```
Comment (OlegHahm) [inline cpu/native/ng_net/dev_eth_tap.c]: remove or don't comment it out

File: cpu/native/ng_net/dev_eth_tap.c [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+#endif
+        }
+
+        _native_in_syscall--;
+
+        return nread;
+    }
+    else if (nread == -1) {
+        if ((errno == EAGAIN) || (errno == EWOULDBLOCK)) {
+            //warn("read");
```
Comment (OlegHahm) [inline cpu/native/ng_net/dev_eth_tap.c]: see above

File: sys/include/dev_eth_autoinit.h [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_net_dev_eth dev_eth auto setup
+ * @ingroup     sys_net_eth
+ * @brief       Automatically setup available ethernet devices
+ * @{
+ *
+ * @file        dev_eth_autoinit.h
```
Comment (OlegHahm) [inline sys/include/dev_eth_autoinit.h]: remove filename

**Comment by kaspar030** [comment]
@OlegHahm I think this is squashable?

**Comment by OlegHahm** [comment]
sure, go ahead

**Comment by kaspar030** [comment]
- squashed
- rebased

File: cpu/native/ng_net/dev_eth_tap.c [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+#endif
+        }
+
+        _native_in_syscall--;
+
+        return nread;
+    }
+    else if (nread == -1) {
+        if ((errno == EAGAIN) || (errno == EWOULDBLOCK)) {
+        }
```
Comment (OlegHahm) [inline cpu/native/ng_net/dev_eth_tap.c]: ?

File: cpu/native/startup.c [hunk 9]
```diff
@@ -367,6 +372,12 @@ __attribute__((constructor)) static void startup(int argc, char **argv)
 #ifdef MODULE_NATIVENET
     tap_init(argv[1]);
 #endif
+#ifdef MODULE_NG_NATIVENET
+# ifdef MODULE_NATIVENET
+#  error  "Modules nativenet and ng_native are mutually exclusive."
+# endif
```
Comment (OlegHahm) [inline cpu/native/startup.c]: Shouldn't we check this in the Makefile?

**Comment by OlegHahm** [comment]
The last two comments are no stoppers. I leave it to you, @kaspar030, if you want to address them. Otherwise go ahead with merging.

File: sys/include/dev_eth_autoinit.h [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_net_dev_eth dev_eth auto setup
+ * @ingroup     sys_net_eth
+ * @brief       Automatically setup available ethernet devices
+ * @{
```
Comment (OlegHahm) [inline sys/include/dev_eth_autoinit.h]: `@file` is missing

File: sys/include/net/dev_eth.h [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for
+ * more details.
+ */
+
+/**
+ * @defgroup    net_dev_eth_ll Low-Level Driver Inteface
+ * @ingroup     net_ng_ethernet
+ * @brief       Low-level ethernet driver interface
```
Comment (OlegHahm) [inline sys/include/net/dev_eth.h]: `@file` is missing

File: cpu/native/ng_net/dev_eth_tap.c (hunk 8 above)
Comment (kaspar030) [inline cpu/native/ng_net/dev_eth_tap.c]: empty if case. :) you made me remove the warning.

**Comment by kaspar030** [comment]
-rebased
-addressed @OlegHahm's doxygen comments

**Comment by kaspar030** [comment]
go, travis, go!

File: cpu/native/ng_net/dev_eth_tap.c (hunk 8 above)
Comment (OlegHahm) [inline cpu/native/ng_net/dev_eth_tap.c]: pffff

**Comment by authmillenon** [comment]
`driver_nrfmin` is failing, but this is unrelated to this PR. Go as soon as it goes … red

**Comment by kaspar030** [comment]
GO.

**Comment by OlegHahm** [comment]
:party: :beers:
